_posts/2023-06-22-a-simple-rest-api-with-play-framework-and-scala.md

```diff
@@ -0,0 +1,420 @@
+import play.api.mvc.{ActionBuilder, ActionTransformer, AnyContent, BodyParser, BodyParsers, PlayBodyParsers, Request, WrappedRequest}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait CarActions {
+  def playBodyParsers: PlayBodyParsers
+  def carManager: CarManager
+  implicit val ec: ExecutionContext
+
+  class CarRequest[A](val carDTO: Option[CarDTO], request: Request[A]) extends WrappedRequest[A](request)
+
+  def CarAction(id: Long) = new ActionBuilder[CarRequest, AnyContent]
+      with ActionTransformer[Request, CarRequest] {
+    def transform[A](request: Request[A]): Future[CarRequest[A]] = {
+      Future{
+        carManager.get(id).fold(_ => new CarRequest(None, request), car => new CarRequest(car, request))
+      }(executionContext)
+    }
+
+    override def parser: BodyParser[AnyContent] = playBodyParsers.default
+
+    override protected def executionContext: ExecutionContext = ec
+  }
+}
+```
+
+This is especially useful in case we need to refine the request to our application. In this specific example we create a [WrappedRequest](https://www.playframework.com/documentation/2.8.x/api/scala/play/api/mvc/WrappedRequest.html) to add the entity information on our Request when a user passes the id on a call.
+We use the [ActionBuilder](https://www.playframework.com/documentation/2.8.x/api/scala/play/api/mvc/ActionBuilder.html) to create an Action that takes a CarRequest instead of a Request. We mix in the [ActionTransformer](https://www.playframework.com/documentation/2.8.x/api/scala/play/api/mvc/ActionTransformer.html) so we can transform the Request object to a CarRequest by doing a call to our manager layer and retrieving a car DTO.
+
+We also need to describe the correspondence between URIs and function calls on our controller. This is achieved through the routes file:
+![alt "Routes structure"](../images/play-rest-api/routes_structure.jpeg)
+
+```scala
+GET     /                 controllers.HomeController.index()
+GET     /car/:id          controllers.HomeController.get(id: Long)
+GET     /cars             controllers.HomeController.getAll()
+
++ nocsrf
+DELETE  /car/delete/:id    controllers.HomeController.delete(id: Long)
++ nocsrf
+PUT     /car/update        controllers.HomeController.update()
++ nocsrf
+POST    /car/save          controllers.HomeController.save()
+```
+Here we define the HTTP method for a specific URI and the method from our controllers layer that will get called when the URI is hit.
+
+In general the Action components of Play are a very flexible way to cover from the default functionality to more advanced.
+
+## 3.4 Application Loader
+Finally we need to put all of the above together and define our main application entry.
+![alt "Application Loader"](../images/play-rest-api/application_loader.jpeg)
+
+Here we define and initialize our main entry point. Initialize all needed components like Managers, Repositories, Routers etc. and finally return an instance of our application. 
```

Comment:
Need to explain the structure of the application - "after I instantiate the components, how do I hook them together?" is the question we need to answer very carefully.